api/client/client.go

```diff
@@ -49,156 +50,293 @@ func init() {
-	if err := cfg.CheckAndSetDefaults(); err != nil {
+// New creates a new API client with an open connection to a Teleport server.
+//
+// The server can be an auth server, web proxy, or tunnel proxy. Multiple server
+// addresses can be specified in cfg.
+//
+// TLS credentials are required in cfg, and can be generated by a number of
+// convenience functions that pull certificates that can be easily generated with tsh.
+// SSH credentials are required to connect to web/tunnel proxies and can only be
+// provided through the identity file credential loader.
+//
+//     LoadIdentityFile("identity-file-path")
+//
+// New will try to open a connection with all combinations of addresses,
+// server types, and credentials. The first successful connection to a server
+// will be used, or an aggregated error will be returned if all combinations fail.
+//
+// If cfg.DialInBackground is true, New will only use the first credentials listed.
+// A predefined dialer must be provided in cfg, or the first addr must be to an auth server.
+// The connection will be dialed in the background, so the connection is not guaranteed
+// to be open. This option is primarily meant for internal use where the client has
+// direct access to server values that guarentee a successful connection.
+func New(ctx context.Context, cfg Config) (clt *Client, err error) {
+	if err = cfg.CheckAndSetDefaults(); err != nil {
 		return nil, trace.Wrap(err)
 	}
 
-	c := &Client{c: cfg}
-	if err := c.connect(ctx); err != nil {
-		return nil, trace.Wrap(err)
+	if cfg.DialInBackground {
+		clt, err = connectInBackground(ctx, cfg)
+	} else {
+		clt, err = connect(ctx, cfg)
 	}
 
-	return c, nil
-}
-
-// GetConnection returns GRPC connection
-func (c *Client) GetConnection() *grpc.ClientConn {
-	return c.conn
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return clt, nil
 }
 
-// getDialer builds a grpc dialer for the client from a ContextDialer.
-// The ContextDialer is chosen from available options, preferring one from
-// credentials, then from configuration, and lastly from addresses.
-func (c *Client) setDialer(creds Credentials) error {
-	var err error
-	if c.dialer, err = creds.Dialer(); err == nil {
-		return nil
+// connectInBackground connects the client to the server in the background.
+// The client will use the first credentials and the given dialer. If
+// no dialer is given, the first address will be used. This address must
+// be an auth server address.
+func connectInBackground(ctx context.Context, cfg Config) (*Client, error) {
+	tlsConfig, err := cfg.Credentials[0].TLSConfig()
+	if err != nil {
+		return nil, trace.Wrap(err)
 	}
-	if c.dialer = c.c.Dialer; c.dialer != nil {
-		return nil
+
+	dialer := cfg.Dialer
+	addr := constants.APIDomain
+	if dialer == nil {
+		if len(cfg.Addrs) == 0 {
+			return nil, trace.BadParameter("must have a Dialer or Addrs in config")
+		}
+		dialer = NewDialer(cfg.KeepAlivePeriod, cfg.DialTimeout)
+		addr = cfg.Addrs[0]
 	}
```

Comment:
how about this, for readability:
```suggestion
	if dialer == nil && len(cfg.Addrs) == 0 {
		return nil, trace.BadParameter("must have a Dialer or Addrs in config")
	}
	if dialer == nil {
		dialer = NewDialer(cfg.KeepAlivePeriod, cfg.DialTimeout)
	}
	if len(cfg.Addrs) > 0 {
		addr = cfg.Addrs[0]
	}
```